pkg/controller/job/backoff_utils.go

```diff
@@ -183,12 +183,18 @@ func getFinishedTime(p *v1.Pod) time.Time {
 }
 
 func getFinishTimeFromContainers(p *v1.Pod) *time.Time {
-	var finishTime *time.Time
-	for _, containerState := range p.Status.ContainerStatuses {
-		if containerState.State.Terminated == nil {
-			return nil
-		}
-		if containerState.State.Terminated.FinishedAt.Time.IsZero() {
+	finishTime := latestFinishTime(nil, p.Status.ContainerStatuses)
+	// We need to check InitContainerStatuses here also,
+	// because with the sidecar (restartable init) containers,
+	// sidecar containers will always finish later than regular containers.
+	return latestFinishTime(finishTime, p.Status.InitContainerStatuses)
```

Comment:
I would suggest to only check finish time of restartable init containers - otherwise we may change the current behavior for some Pod that behaves badly and reports incorrect finish time for the regular init container. I am mostly concerned of some badly written controllers that set this time to some random large value for the regular init containers.

Since sidecars are new, accounting for them is not supposed to break annything.

Reply:
> I am mostly concerned of some badly written controllers that set this time to some random large value for the regular init containers.

Well, I'm not aware of this... 
I added check in the second commit, let's wait for some reviews.

Reply:
I liked the simplicity of the previous solution, but I'm not very familiar with the sidecar containers yet.

> I am mostly concerned of some badly written controllers that set this time to some random large value for the regular init containers.

Would the container finish time, in case of sidecar containers, be set by a user-supplied controller? I think if the finish at time is set by kubelet then we could rely on this. 

cc @alculquicondor

Reply:
It is set by kubelet

Reply:
I think we can simplify the code and check the InitContainers regardless of the `restartPolicy`. For safety, we just wrap with the `SidecarContainers` feature-gate check as suggested in https://github.com/kubernetes/kubernetes/pull/124942#discussion_r1608737775. 

This way we complicate the code only until the feature-gate is around. WDYT @alculquicondor @SergeyKanzhelev ?

Reply:
My only worry is to encounter some crazy controller like one that made me write this comment: https://github.com/kubernetes/kubernetes/pull/124918. Semantically I am all for not checking for `restartPolicy`. Breaking some barely supported scenario is what keeping me uneasy here

Reply:
In that case, I agree with the current implementation that restricts the change to initContainers with restartPolicy: Always.